tests/src/test/scala/org/apache/openwhisk/core/containerpool/test/ContainerPoolTests.scala

```diff
@@ -753,14 +758,211 @@ class ContainerPoolTests
+            List(PrewarmingConfig(initialCount, exec, memoryLimit, reactive))))
+    containers(0).expectMsg(Start(exec, memoryLimit, Some(ttl)))
+    containers(1).expectMsg(Start(exec, memoryLimit, Some(ttl)))
+    containers(0).send(pool, NeedWork(preWarmedData(exec.kind, expires = deadline)))
+    containers(1).send(pool, NeedWork(preWarmedData(exec.kind, expires = deadline)))
+
+    // when invoker starts, include 0 prewarm container at the very beginning
+    stream.toString should include(s"currentCount: 0 prewarmed container")
+
+    // the desiredCount should equal with initialCount when invoker starts
+    stream.toString should include(s"needs ${initialCount} desired prewarmed container")
+
+    stream.reset()
+
+    // Make sure the created prewarmed containers are expired
+    Thread.sleep(ttl.toMillis)
+    pool ! AdjustPrewarmedContainer
+    // Make sure adjustPrewarmContainer run finished
+    Thread.sleep(2.seconds.toMillis)
+
+    containers(0).expectMsg(Remove)
+    containers(1).expectMsg(Remove)
+    containers(0).send(pool, ContainerRemoved(false))
+    containers(1).send(pool, ContainerRemoved(false))
+
+    // Because already supplemented the prewarmed container, so currentCount should equal with initialCount
+    stream.toString should include(s"currentCount: ${initialCount} prewarmed container")
+
+    // the desiredCount should equal with minCount because cold start didn't happen
+    stream.toString should include(s"needs ${minCount} desired prewarmed container")
+    // Previously created prewarmed containers should be removed
+    stream.toString should include(s"removed ${initialCount} expired prewarmed container")
+
+    stream.reset()
+    val action = ExecutableWhiskAction(
+      EntityPath("actionSpace"),
+      EntityName("actionName"),
+      exec,
+      limits = ActionLimits(memory = MemoryLimit(memoryLimit)))
+    val run = createRunMessage(action, invocationNamespace)
+    // 2 code start happened
+    pool ! run
+    pool ! run
+    containers(2).expectMsg(run)
+    containers(3).expectMsg(run)
+
+    pool ! AdjustPrewarmedContainer
+    // Make sure adjustPrewarmContainer run finished
+    Thread.sleep(2.seconds.toMillis)
+
+    // Because already removed expired prewarmed containrs, so currentCount should equal with 0
+    stream.toString should include(s"currentCount: 0 prewarmed container")
+    // the desiredCount should equal with 2 due to cold start happened
+    stream.toString should include(s"needs 2 desired prewarmed container")
+
+    containers(4).expectMsg(Start(exec, memoryLimit, Some(ttl)))
+    containers(5).expectMsg(Start(exec, memoryLimit, Some(ttl)))
+    containers(4).send(pool, NeedWork(preWarmedData(exec.kind, expires = deadline)))
+    containers(5).send(pool, NeedWork(preWarmedData(exec.kind, expires = deadline)))
+
+    // Make sure the created prewarmed containers are expired
+    stream.reset()
+    Thread.sleep(ttl.toMillis)
+    pool ! AdjustPrewarmedContainer
```

Comment:
shouldn't pool send `AdjustPrewarmedContainer` to itself eventually after the `poolConfig.prewarmExpiredCheckPeriod`?

Reply:
I think need, bacause the prewarmExpirationCheckInterval is `1 minute` 
```
  def poolConfig(userMemory: ByteSize) =
    ContainerPoolConfig(userMemory, 0.5, false, FiniteDuration(1, TimeUnit.MINUTES))
```
so need to send AdjustPrewarmedContainer  to itself manually in codes.
Why i don't change above value to a less value, the reason is `i want to make user's to know the detail clearly when see the test codes`

Reply:
But the test is written in a way suggesting that the `AdjustPrewarmedContainer ` message is explicitly sent by logic, but it never is - it is only ever sent via schedule. So it would be more clear to rely on the timing (with a shorter value) and see that the results of the message occur without interference of the test logic.

Reply:
Updated accordingly.